### PR TITLE
Add support for more process types

### LINKED ISSUE #28: Improve Procfile parsing
The `Procfile` parsing is very basic at the moment. We should look for and use other process types besides "web" if web is not found. If more than a single process type is found, we should also warn/error to provide feedback to the user.

### PR DESCRIPTION
**Pull request by wyzlle**
Closes #28

### CODE REVIEW

**Comment by wyzlle** [comment]
Should I implement support for the `release` process type?

File: src/nixpacks/plan/generator.rs [hunk 1]
```diff
@@ -231,16 +239,14 @@ impl<'a> NixpacksBuildPlanGenerator<'a> {
...
-            let contents = app.read_file("Procfile")?;
-
-            // Better error handling
-            if contents.starts_with("web: ") {
-                return Ok(Some(contents.replace("web: ", "").trim().to_string()));
+            let procfile: Procfile = app.read_yaml("Procfile").unwrap_or_default();
+            if procfile.web.is_some() && procfile.worker.is_some() {
+                bail!("Procfile contains both a worker and a web process. Please specify one.");
+            } else {
+                Ok(procfile.web.or(procfile.worker))
```
Comment (jozsefsallai) [inline src/nixpacks/plan/generator.rs:248]: Maybe we should also warn the user if neither of the two are found? e.g. if they have something weird going on like
```yaml
web.1: npm run first
web.2: npm run second
```
In this scenario, the user doesn't technically have any of the supported process types. Would be cool to let them know why the build plan still runs the default commands.

**Review by jozsefsallai — CHANGES_REQUESTED** [review]

File: src/nixpacks/plan/generator.rs [hunk 2]
```diff
@@ -32,6 +33,13 @@ pub struct NixpacksBuildPlanGenerator<'a> {
     options: GeneratePlanOptions,
 }
 
+#[derive(Deserialize, Default)]
+pub struct Procfile {
+    pub web: Option<String>,
+    pub worker: Option<String>,
+    // pub release: Option<String>,
```
[the remark this answers is not in the document]
Comment (wyzlle) [reply src/nixpacks/plan/generator.rs:40]: Yeah, will do that. Just wanted to confirm. [link](https://github.com/railwayapp/nixpacks/pull/181#issuecomment-1146697841)

File: src/nixpacks/plan/generator.rs [hunk 3]
```diff
@@ -231,16 +239,14 @@ impl<'a> NixpacksBuildPlanGenerator<'a> {
...
 
-    fn parse_procfile(&self, app: &App) -> Result<Option<String>> {
+    fn get_procfile_start_cmd(&self, app: &App) -> Result<Option<String>> {
         if app.includes_file("Procfile") {
-            let contents = app.read_file("Procfile")?;
-
-            // Better error handling
-            if contents.starts_with("web: ") {
-                return Ok(Some(contents.replace("web: ", "").trim().to_string()));
+            let procfile: Procfile = app.read_yaml("Procfile").unwrap_or_default();
```
[the remark this answers is not in the document]
Comment (coffee-cup) [reply src/nixpacks/plan/generator.rs:244]: I don't think the performance of parsing of a relatively small yaml file should be of any concern. Serde is fairly fast to parse anyway and I can't imagine it is any more than a millisecond difference.

File: src/nixpacks/plan/generator.rs (hunk 1 above)
Comment (coffee-cup) [reply src/nixpacks/plan/generator.rs:248]: +1 that is a good idea

**Review by coffee-cup — APPROVED** [review]
Changes look good pending a few minor comments.

File: src/nixpacks/plan/generator.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jozsefsallai) [reply src/nixpacks/plan/generator.rs:244]: Fair enough, was just wondering.

File: src/nixpacks/plan/generator.rs [hunk 4]
```diff
@@ -165,11 +165,20 @@ impl<'a> NixpacksBuildPlanGenerator<'a> {
             .or(env_build_cmd)
             .or(build_phase.cmd);
 
+        // Release process type
+        if let Some(release_cmd) = self.get_procfile_release_cmd(app)? {
```
Comment (coffee-cup) [inline src/nixpacks/plan/generator.rs:169]: I may be missing something, but why is the release command being used as the build command?

**Review by coffee-cup — APPROVED** [review]
Looks good besides the one question

File: src/nixpacks/plan/generator.rs (hunk 4 above)
Comment (wyzlle) [reply src/nixpacks/plan/generator.rs:169]: Release is run after build but before start ([source](https://devcenter.heroku.com/articles/release-phase)). And Cooper was [against adding](https://canary.discord.com/channels/713503345364697088/982408838063988756) new phases.

**Pull request merged**